src/KubernetesClient/Informers/Cache/VersionPartitionedSharedCache.cs

```diff
@@ -0,0 +1,294 @@
+        {
+            _keySelector = keySelector;
+            _versionSelector = versionSelector;
+        }
+
+        /// <summary>
+        /// Creates a unique cache partition that may share references to objects with same key/versions with other partitions
+        /// </summary>
+        /// <returns>Partitioned cache</returns>
+        public ICache<TKey, TResource> CreatePartition()
+        {
+            lock (_lock)
+            {
+                var view = new CacheView(this);
+                _views.Add(view);
+                return view;
+            }
+        }
+
+
+        private void Remove(TResource resource, CacheView originView)
+        {
+            var versionedKey = GetVersionKeyFor(resource);
+            Remove(versionedKey, originView);
+        }
+
+        private void Remove(VersionResourceKey versionedKey, CacheView originView)
+        {
+            var otherViewsTrackingResource = _views
+                .Except(new[] {originView})
+                .Any(x => x.TryGetValue(versionedKey.Key, out var resource) && _versionSelector(resource).Equals(versionedKey.Version));
+            if (!otherViewsTrackingResource)
+            {
+                Items.Remove(versionedKey);
+            }
+        }
+
+        private TResource GetOrAdd(TResource resource)
+        {
+            var key = GetVersionKeyFor(resource);
+            if (Items.TryGetValue(key, out var existingResource))
+                return existingResource;
+            Items.Add(key, resource);
+            return resource;
+        }
+
+        internal struct VersionResourceKey
+        {
+            public TKey Key;
+            public TVersion Version;
+        }
+
+        private VersionResourceKey GetVersionKeyFor(TResource resource)
+        {
+            return new VersionResourceKey() {Key = _keySelector(resource), Version = _versionSelector(resource)};
+        }
+
+        private class CacheView : ICache<TKey, TResource>
+        {
+            private readonly VersionPartitionedSharedCache<TKey, TResource, TVersion> _parent;
+            private readonly Dictionary<TKey, TResource> _items = new Dictionary<TKey, TResource>() ;
+
+            public long Version { get; set; }// = 1;
+
+            public CacheView(VersionPartitionedSharedCache<TKey, TResource, TVersion> parent)
+            {
+                _parent = parent;
+            }
+
+            public void Reset(IDictionary<TKey, TResource> newValues)
+            {
+                lock (_parent._lock)
+                {
+                    _items.Clear();
+                    foreach (var item in newValues)
+                    {
+                        _items.Add(item.Key, item.Value);
+                    }
+                }
+            }
+
+            public object SyncRoot => _parent._lock;
+            public ICacheSnapshot<TKey, TResource> Snapshot()
+            {
+                lock (_parent)
```

Comment:
Shouldn't this be `_parent._lock`? (if not, document why not)

Reply:
Also, why are you locking around a constructor? Shouldn't that be thread-safe?

Reply:
1. Bug: changed to _parent._lock
2. The lock is there because snapshot uses current object as a  source dictionary (which gets copied in the constructor). This copying is not thread safe.